File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-0743-1.cpp
class Solution {
public:
    /// Simply run Dijkra's shortest path algorithm & take the distance
    /// to the farthest node as the solution
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        // Convert times to an adjacency list
        vector<vector<pair<int, int>>> G(n+1);
        for (int i=0; i<times.size(); i++)
            G[times[i][0]].push_back({times[i][1], times[i][2]});

        // Run Dijktra's algorithm
        enum Color {BLACK, WHITE};
        vector<Color> color(n+1, WHITE);
        vector<int> cost(n+1, INT_MAX);
        cost[k] = 0;

        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       std::greater<pair<int, int>>> Q;
        Q.push({0, k});
        
        while (Q.size() > 0){
            // Find the node with the min distance
            auto [currCost, u] = Q.top(); Q.pop();
            if (color[u] == BLACK) continue;

            color[u] = BLACK;  // Mark as processed
                
            for (auto [v, w]: G[u]){
                if (color[v] == BLACK) continue;
                    
                int newCost = currCost + w;
                if (newCost < cost[v]){
                    cost[v] = newCost;
                    Q.push({newCost, v});                    
                } // end-if
            } // end-for
        } // end-while

        int maxVal = *std::max_element(cost.begin()+1, cost.end());        
        return maxVal == INT_MAX? -1: maxVal;
    } // end-networkDelayTime
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-0743-1.py
import heapq
class Solution:
    # Simply run Dijkra's shortest path algorithm & take the distance
    # to the farthest node as the solution
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Convert times to an adjacency list
        G = []
        for i in range(n+1): G.append([])
            
        for i in range(len(times)):
            G[times[i][0]].append([times[i][1], times[i][2]])

        # Run Dijktra's algorithm
        BLACK = 0
        WHITE = 2
        color = [WHITE]*(n+1)
        cost = [20000]*(n+1)

        cost[0] = 0      # There is no node with id=0. So, make its distance 0
        cost[k] = 0
        Q = [(0, k)]
        
        while (len(Q) > 0):
            # Get the node with the smallest distance
            currCost, u = heapq.heappop(Q)
            
            if color[u] == BLACK: continue  # Already processed
            
            color[u] = BLACK # Mark as processed

            for v, w in G[u]:            
                if (color[v] == BLACK): continue
                    
                newCost = currCost + w
                if (newCost < cost[v]):                    
                    cost[v] = newCost
                    heapq.heappush(Q, [newCost, v])

        for u in range(1, n+1):                    
            if (color[u] == WHITE):
                return -1
        
        return max(cost)        
        
        